File: days/day04.py
from etc.utils import file_to_lines
import re

passports_raw = file_to_lines("input/day04.txt", spl="\n\n")
RE_PASS = re.compile(r"""([a-z]{3}):(.*?)(?:\s|$)""")

def line_to_passport(line):
    return dict(m.groups() for m in RE_PASS.finditer(line))
passports = list(map(line_to_passport, passports_raw))

regexs = {
    "byr": re.compile(r"""^(19[2-9]\d|200[0-2])$"""),
    "iyr": re.compile(r"""^(201\d|2020)$"""),
    "eyr": re.compile(r"""^(202\d|2030)$"""),
    "hgt": re.compile(r"""^(1([5-8]\d|9[0-3])cm|(59|6\d|7[0-6])in)$"""),
    "hcl": re.compile(r"""^#[0-9a-f]{6}$"""),
    "ecl": re.compile(r"""^(amb|blu|brn|gry|grn|hzl|oth)$"""),
    "pid": re.compile(r"""^[0-9]{9}$"""),
}

# Parts 1 & 2
is_valid_passport_1 = lambda pp: all(f in pp for f in regexs)
is_valid_passport_2 = lambda pp: all(rgx.match(pp[k]) for k, rgx in regexs.items())  # Should only be applied to passports that meet the previous criterion

valid_1 = valid_2 = 0

for pp in passports:
    if is_valid_passport_1(pp):
        valid_1 += 1
        if is_valid_passport_2(pp):
            valid_2 += 1

print("Part 1:", valid_1)
print("Part 2:", valid_2)

File: rust/src/days/day04.rs
use std::io::{BufReader, BufRead};
use std::fs::File;
use std::time::Instant;
use regex::Regex;

const RE_LIN: &str = r"([a-z]{3}):(.*?)(?:\s|$)";
const RE_BYR: &str = r"^(19[2-9]\d|200[0-2])$";
const RE_IYR: &str = r"^(201\d|2020)$";
const RE_EYR: &str = r"^(202\d|2030)$";
const RE_HGT: &str = r"^(1([5-8]\d|9[0-3])cm|(59|6\d|7[0-6])in)$";
const RE_HCL: &str = r"^#[0-9a-f]{6}$";
const RE_ECL: &str = r"^(amb|blu|brn|gry|grn|hzl|oth)$";
const RE_PID: &str = r"^[0-9]{9}$";

///////////////////////////////////////////////////////////////////////////////
// This one could be much more efficient... I'll get back to it some day, maybe

pub fn run() {
    let time = Instant::now();

    let f = BufReader::new(File::open("../input/day04.txt").unwrap());
    let mut sol_part_1 = 0;
    let mut sol_part_2 = 0;

    let mut acc1 = 0;
    let mut acc2 = 0;

    let regex_lin = Regex::new(RE_LIN).unwrap();
    let regex_byr = Regex::new(RE_BYR).unwrap();
    let regex_iyr = Regex::new(RE_IYR).unwrap();
    let regex_eyr = Regex::new(RE_EYR).unwrap();
    let regex_hgt = Regex::new(RE_HGT).unwrap();
    let regex_hcl = Regex::new(RE_HCL).unwrap();
    let regex_ecl = Regex::new(RE_ECL).unwrap();
    let regex_pid = Regex::new(RE_PID).unwrap();

    for line in f.lines() {
        let line = line.unwrap();
        if line == "" {
            if acc1 == 7 { sol_part_1 += 1; }
            if acc2 == 7 { sol_part_2 += 1; }
            acc1 = 0;
            acc2 = 0;
        }

        for cap in regex_lin.captures_iter(&line) {
            let (key, val) = (cap.get(1).unwrap().as_str(), cap.get(2).unwrap().as_str());
            let re = match key {
                "byr" => &regex_byr,
                "iyr" => &regex_iyr,
                "eyr" => &regex_eyr,
                "hgt" => &regex_hgt,
                "hcl" => &regex_hcl,
                "ecl" => &regex_ecl,
                "pid" => &regex_pid,
                "cid" => continue,
                _ => panic!("unknown key"),
            };

            acc1 += 1;
            if re.is_match(val) {
                acc2 += 1;
            }
        }
    }

    if acc1 == 7 { sol_part_1 += 1; }
    if acc2 == 7 { sol_part_2 += 1; }

    let elapsed_ms = time.elapsed().as_nanos() as f64 / 1_000_000.0;
    println!("Part 1: {}", sol_part_1);
    println!("Part 2: {}", sol_part_2);
    println!("Elapsed: {:.3} ms", elapsed_ms);
}
